File: backend/capsAllocation/src/main/java/com/vbs/capsAllocation/service/DropdownConfigurationService.java
package com.vbs.capsAllocation.service;

import com.vbs.capsAllocation.dto.CreateDropdownConfigurationDTO;
import com.vbs.capsAllocation.dto.DropdownConfigurationDTO;
import com.vbs.capsAllocation.dto.UpdateDropdownConfigurationDTO;

import java.util.List;

/**
 * Service interface for managing dropdown configurations
 */
public interface DropdownConfigurationService {

    /**
     * Create a new dropdown configuration option
     * @param createDTO The data for creating the dropdown option
     * @param createdBy The username of the user creating the option
     * @return The created dropdown configuration
     */
    DropdownConfigurationDTO createDropdownOption(CreateDropdownConfigurationDTO createDTO, String createdBy);

    /**
     * Get all active dropdown options for a specific type
     * @param dropdownType The type of dropdown
     * @return List of active dropdown configurations
     */
    List<DropdownConfigurationDTO> getActiveDropdownOptions(String dropdownType);

    /**
     * Get all dropdown options for a specific type (including inactive)
     * @param dropdownType The type of dropdown
     * @return List of all dropdown configurations
     */
    List<DropdownConfigurationDTO> getAllDropdownOptions(String dropdownType);

    /**
     * Update an existing dropdown configuration option
     * @param id The ID of the dropdown option to update
     * @param updateDTO The data for updating the dropdown option
     * @return The updated dropdown configuration
     */
    DropdownConfigurationDTO updateDropdownOption(Long id, UpdateDropdownConfigurationDTO updateDTO);

    /**
     * Delete a dropdown configuration option
     * @param id The ID of the dropdown option to delete
     */
    void deleteDropdownOption(Long id);

    /**
     * Get a specific dropdown option by ID
     * @param id The ID of the dropdown option
     * @return The dropdown configuration
     */
    DropdownConfigurationDTO getDropdownOptionById(Long id);

    /**
     * Get all dropdown types
     * @return List of distinct dropdown types
     */
    List<String> getAllDropdownTypes();

    /**
     * Reorder dropdown options for a specific type
     * @param dropdownType The type of dropdown
     * @param orderedIds List of IDs in the desired order
     * @return List of reordered dropdown configurations
     */
    List<DropdownConfigurationDTO> reorderDropdownOptions(String dropdownType, List<Long> orderedIds);

    /**
     * Check if user has permission to manage dropdown configurations
     * @param userRole The role of the current user
     * @return true if user can manage dropdowns, false otherwise
     */
    boolean canManageDropdowns(String userRole);
}

File: org-chart/src/app/services/dropdown-configuration.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { map } from 'rxjs/operators';
import { environment } from '../../environments/environment';
import {
 DropdownConfiguration,
 CreateDropdownConfiguration,
 UpdateDropdownConfiguration,
 BaseResponse
} from '../model/dropdown-configuration.interface';


@Injectable({
 providedIn: 'root'
})
export class DropdownConfigurationService {
 private baseUrl = `${environment.apiUrl}/api/dropdown-configurations`;


 constructor(private http: HttpClient) {}


 /**
  * Create a new dropdown configuration option
  */
 createDropdownOption(createData: CreateDropdownConfiguration): Observable<DropdownConfiguration> {
   return this.http.post<BaseResponse<DropdownConfiguration>>(this.baseUrl, createData).pipe(
     map(response => response.data)
   );
 }


 /**
  * Get all active dropdown options for a specific type
  */
 getActiveDropdownOptions(dropdownType: string): Observable<DropdownConfiguration[]> {
   return this.http.get<BaseResponse<DropdownConfiguration[]>>(`${this.baseUrl}/active/${dropdownType}`).pipe(
     map(response => response.data)
   );
 }


 /**
  * Get all dropdown options for a specific type (including inactive)
  */
 getAllDropdownOptions(dropdownType: string): Observable<DropdownConfiguration[]> {
   return this.http.get<BaseResponse<DropdownConfiguration[]>>(`${this.baseUrl}/all/${dropdownType}`).pipe(
     map(response => response.data)
   );
 }


 /**
  * Update an existing dropdown configuration option
  */
 updateDropdownOption(id: number, updateData: UpdateDropdownConfiguration): Observable<DropdownConfiguration> {
   return this.http.put<BaseResponse<DropdownConfiguration>>(`${this.baseUrl}/${id}`, updateData).pipe(
     map(response => response.data)
   );
 }


 /**
  * Delete a dropdown configuration option
  */
 deleteDropdownOption(id: number): Observable<void> {
   return this.http.delete<BaseResponse<void>>(`${this.baseUrl}/${id}`).pipe(
     map(() => void 0)
   );
 }


 /**
  * Get a specific dropdown option by ID
  */
 getDropdownOptionById(id: number): Observable<DropdownConfiguration> {
   return this.http.get<BaseResponse<DropdownConfiguration>>(`${this.baseUrl}/${id}`).pipe(
     map(response => response.data)
   );
 }


 /**
  * Get all dropdown types
  */
 getAllDropdownTypes(): Observable<string[]> {
   return this.http.get<BaseResponse<string[]>>(`${this.baseUrl}/types`).pipe(
     map(response => response.data)
   );
 }


 /**
  * Reorder dropdown options for a specific type
  */
 reorderDropdownOptions(dropdownType: string, orderedIds: number[]): Observable<DropdownConfiguration[]> {
   return this.http.put<BaseResponse<DropdownConfiguration[]>>(`${this.baseUrl}/reorder/${dropdownType}`, orderedIds).pipe(
     map(response => response.data)
   );
 }


 /**
  * Check if user has admin/ops manager role
  */
 canManageDropdowns(): boolean {
   const role = localStorage.getItem('role');
   return role === 'ADMIN_OPS_MANAGER';
 }
}



